spharpy/spherical.py

```diff
@@ -1005,3 +1007,413 @@ def sph_identity_matrix(n_max, type='n-nm'):
+            self._compute_basis()
+            self._recompute_basis = False
+        return self._basis
+
+    @property
+    def basis_gradient_theta(self):
+        """ndarray: Get the gradient of the basis matrix with
+        respect to theta."""
+        if (self._basis_gradient_theta is None or
+                self._recompute_basis_gradient):
+            self._compute_basis_gradient()
+            self._recompute_basis_gradient = False
+        return self._basis_gradient_theta
+
+    @property
+    def basis_gradient_phi(self):
+        """ndarray: Get the gradient of the basis
+        matrix with respect to phi."""
+        if (self._basis_gradient_phi is None or
+                self._recompute_basis_gradient):
+            self._compute_basis_gradient()
+            self._recompute_basis_gradient = False
+        return self._basis_gradient_phi
+
+    def _compute_basis(self):
+        """
+        Compute the basis matrix for the SphericalHarmonics class.
+        """
+        if self.basis_type == "complex":
+            function = spherical_harmonic_basis
+        elif self.basis_type == "real":
+            function = spherical_harmonic_basis_real
+        else:
+            raise ValueError(
+                "Invalid basis type, should be either 'complex' or 'real'"
+            )
+        self._basis = function(
+            self.n_max, self.coordinates,
+            self.normalization, self.channel_convention)
+
+    def _compute_basis_gradient(self):
+        """
+        Compute the gradient of the basis matrix for the SphericalHarmonics class
+        """
+        if any(
+            (self.normalization in ["maxN", "sn3d"],
+             self.channel_convention == "fuma")
+        ):
+            raise ValueError(
+                "Gradient computation not supported for MaxN, "
+                "SN3D normalization or FuMa channel ordering"
+            )
+        else:
+            if self.basis_type == "complex":
+                function = spherical_harmonic_basis_gradient
+            elif self.basis_type == "real":
+                function = spherical_harmonic_basis_gradient_real
+            else:
+                raise ValueError(
+                    "Invalid basis type, should be either 'complex' or 'real'"
+                )
+            self._basis_gradient_theta, self._basis_gradient_phi = function(
+                self.n_max, self.coordinates)
+
+    @property
+    def basis_inv(self):
+        """ndarray: Get the inverse basis matrix."""
+        if (self._basis_inv is None or
+                self._recompute_inverse):
+            self._compute_inverse()
+            self._recompute_inverse = False
+        return self._basis_inv
+
+    def _compute_inverse(self):
+        """
+        Compute the inverse basis matrix for the SphericalHarmonics class
+        """
+        if self._basis is None:
+            self._compute_basis()
+        if self.inverse_transform is not None:
```

Comment:
See comment in `__init__` maybe we do not initialize this with `None` and than also dont need to check for it here.